File: Rnwood.Smtp4dev/ClientApp/src/ApiClient/UpdatesController.ts
import axios from "axios";
import UpdateCheckResult from "./UpdateCheckResult";

export default class UpdatesController {
    constructor() {}

    public checkForUpdates_url(username: string | null = null): string {
        let url = "api/Updates/check";
        if (username) {
            url += `?username=${encodeURIComponent(username)}`;
        }
        return url;
    }

    public async checkForUpdates(username: string | null = null): Promise<UpdateCheckResult> {
        return (await axios.get(this.checkForUpdates_url(username))).data as UpdateCheckResult;
    }

    public markVersionAsSeen_url(username: string, version: string): string {
        return `api/Updates/mark-seen?username=${encodeURIComponent(username)}&version=${encodeURIComponent(version)}`;
    }

    public async markVersionAsSeen(username: string, version: string): Promise<void> {
        await axios.post(this.markVersionAsSeen_url(username, version));
    }
}

File: Rnwood.Smtp4dev/Controllers/UpdatesController.cs
using System;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Rnwood.Smtp4dev.ApiModel;
using Rnwood.Smtp4dev.Service;

namespace Rnwood.Smtp4dev.Controllers
{
    /// <summary>
    /// Handles update notifications and what's new functionality
    /// </summary>
    [Route("api/[controller]")]
    [ApiController]
    public class UpdatesController : Controller
    {
        private readonly UpdateNotificationService _updateService;

        public UpdatesController(UpdateNotificationService updateService)
        {
            _updateService = updateService;
        }

        /// <summary>
        /// Checks for updates and new releases
        /// </summary>
        /// <param name="username">Optional username to track version per user</param>
        /// <returns>Update check result with available updates and what's new information</returns>
        [HttpGet("check")]
        public async Task<ActionResult<UpdateCheckResult>> CheckForUpdates([FromQuery] string username = null)
        {
            var result = await _updateService.CheckForUpdatesAsync(username);
            return Ok(result);
        }

        /// <summary>
        /// Marks a version as seen by the user
        /// </summary>
        /// <param name="username">Username to track version for</param>
        /// <param name="version">Version that was seen</param>
        [HttpPost("mark-seen")]
        public async Task<ActionResult> MarkVersionAsSeen([FromQuery] string username, [FromQuery] string version)
        {
            await _updateService.MarkVersionAsSeenAsync(username ?? "anonymous", version);
            return Ok();
        }
    }
}
